pkg/controller/syncmachineset/syncmachineset_controller.go

```
@@ -0,0 +1,563 @@
+
+	capiMachineSet := clusterapiv1.MachineSet{}
+	capiMachineSet.Name = machineSet.Name
+	capiMachineSet.Namespace = remoteClusterAPINamespace
+
+	// Set replicas = 0 so that we don't create machines yet.
+	// TODO: remove me when ready to create machines from remote.
+	// replicas := int32(machineSet.Spec.Size)
+	replicas := int32(0)
+	capiMachineSet.Spec.Replicas = &replicas
+
+	capiMachineSet.Spec.Selector.MatchLabels = map[string]string{"machineset": machineSet.Name}
+
+	if machineSet.Annotations == nil {
+		machineSet.Annotations = map[string]string{}
+	}
+	machineSet.Annotations["cluster-operator.openshift.io/cluster-version"] = string(serializeCOResource(clusterVersion))
+
+	machineTemplate := clusterapiv1.MachineTemplateSpec{}
+	machineTemplate.Labels = map[string]string{"machineset": machineSet.Name}
+	machineTemplate.Spec.Labels = map[string]string{"machineset": machineSet.Name}
+	machineTemplate.Spec.ProviderConfig.Value = &runtime.RawExtension{
+		Raw: serializeCOResource(machineSet),
+	}
+	capiMachineSet.Spec.Template = machineTemplate
+
+	return &capiMachineSet
+}
+
+func buildClusterAPICluster(cluster *cov1.Cluster) *clusterapiv1.Cluster {
+	capiCluster := clusterapiv1.Cluster{}
+	capiCluster.Name = cluster.Name
+	capiCluster.Spec.ProviderConfig.Value = &runtime.RawExtension{
+		Raw: serializeCOResource(cluster),
+	}
+
+	// These are unused, dummy values.
+	capiCluster.Spec.ClusterNetwork.ServiceDomain = "cluster-api.k8s.io"
+	capiCluster.Spec.ClusterNetwork.Pods.CIDRBlocks = []string{"10.10.0.0/16"}
+	capiCluster.Spec.ClusterNetwork.Services.CIDRBlocks = []string{"172.30.0.0/16"}
+
+	return &capiCluster
+}
+
+func serializeCOResource(object runtime.Object) []byte {
+	serializer := jsonserializer.NewSerializer(jsonserializer.DefaultMetaFactory, coapi.Scheme, coapi.Scheme, false)
+	encoder := coapi.Codecs.EncoderForVersion(serializer, cov1.SchemeGroupVersion)
+	buffer := &bytes.Buffer{}
+	encoder.Encode(object, buffer)
```

Don't swallow errors returning from `Encode`.